File: 2493-reverse-odd-levels-of-binary-tree/2493-reverse-odd-levels-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */

//  Time Complexity: O(N) → Every node is visited once.

// Space Complexity: O(N) → BFS queue + level-wise storage.
var reverseOddLevels = function (root) {
    if (!root) return null;
    let queue = [root];
    let level = 0;
    while (queue.length > 0) {  // O(N) - We process each node once
        let size = queue.length; // O(1) - Get level size
        let levelNodes = []; // O(1) - Store nodes for swapping

        for (let i = 0; i < size; i++) {  // O(N) - Traverse level nodes
            let node = queue.shift();  // O(1) - Remove from queue
            levelNodes.push(node);  // O(1) - Store node for reversing

            if (node.left) queue.push(node.left);  // O(1) - Push left child
            if (node.right) queue.push(node.right);  // O(1) - Push right child
        }

        if (level % 2 === 1) {  // Only reverse on odd levels
            let left = 0, right = levelNodes.length - 1;
            while (left < right) {  // O(K) - Reverse level values
                [levelNodes[left].val, levelNodes[right].val] =
                 [levelNodes[right].val, levelNodes[left].val];
                left++;
                right--;
            }
        }

        level++;  // Move to next level
    }
    return root;
};

File: 2493-reverse-odd-levels-of-binary-tree/2493-reverse-odd-levels-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode reverseOddLevels(TreeNode root) {
       swapHelper(root.left, root.right, 1);
       return root;
    }

private void swapHelper(TreeNode left, TreeNode right, int level) {
    if (left == null || right == null) return;

    if (level % 2 == 1) {
        int temp = left.val;
        left.val = right.val;
        right.val = temp;
    }

    swapHelper(left.left, right.right, level + 1);
    swapHelper(left.right, right.left, level + 1);
}


}